### PR TITLE
Support for In Memory Subscription validation

### PR DESCRIPTION
**Pull request by isharac**
This PR provides the support for In Memory Subscription Validation in the gateway.
New REST APIs are introduced to load meta data from the db.

Ex:
**/apis**
curl --location  --request GET 'https://localhost:9443/internal/data/v1/apis' -kv -header 'X-WSO2-Tenant: abc.com'
{"count":1,"list":[{"apiId":1,"provider":"adminfoo@abc.com","name":null,"version":"1.0.0","context":"/t/abc.com/pizzashack/1.0.0","policy":null,"urlMaping":[]}],"pagination":null}

curl --location  --request GET 'https://localhost:9443/internal/data/v1/apis?vesion=1.0.0&context=/t/abc.com/pizzashack/1.0.0' -kv
{"count":1,"list":[{"apiId":1,"provider":"adminfoo@abc.com","name":null,"version":"1.0.0","context":"/t/abc.com/pizzashack/1.0.0","policy":null,"urlMaping":[]}],"pagination":null}

curl --location  --request GET 'https://localhost:9443/internal/data/v1/apis’
{"count":2,"list":[{"id":3,"name":null,"subName":"adminfoo@abc.com","policy":"Unlimited","tokenType":"JWT","groupIds":[],"attributes":[]},{"id":4,"name":null,"subName":"adminfoo@abc.com","policy":"10PerMin","tokenType":"JWT","groupIds":[],"attributes":[]}],"pagination":null}

**/subscriptions**
curl --location  --request GET 'https://localhost:9443/internal/data/v1/subscriptions'
{"count":1,"list":[{"subscriptionId":1,"policyId":"Unlimited","apiId":1,"appId":4,"subscriptionState":"UNBLOCKED"}],"pagination":null}

curl --location  --request GET 'https://localhost:9443/internal/data/v1/subscriptions?apiId=1&appId=4' -kv
{"count":1,"list":[{"subscriptionId":1,"policyId":"Unlimited","apiId":1,"appId":4,"subscriptionState":"UNBLOCKED"}],"pagination":null}

curl --location  --request GET 'https://localhost:9443/internal/data/v1/subscriptions' -kv --header 'X-WSO2-Tenant: abc.com'
{"count":1,"list":[{"subscriptionId":1,"policyId":"Unlimited","apiId":1,"appId":4,"subscriptionState":"UNBLOCKED"}],"pagination":null}

**/applictaions**
curl --location  --request GET 'https://localhost:9443/internal/data/v1/applications’
{"count":2,"list":[{"id":3,"name":null,"subName":"adminfoo@abc.com","policy":"Unlimited","tokenType":"JWT","groupIds":[],"attributes":[]},{"id":4,"name":null,"subName":"adminfoo@abc.com","policy":"10PerMin","tokenType":"JWT","groupIds":[],"attributes":[]}],"pagination":null}

curl --location  --request GET 'https://localhost:9443/internal/data/v1/applications?appId=4' -kv
{"count":1,"list":[{"id":4,"name":null,"subName":"adminfoo@abc.com","policy":"10PerMin","tokenType":"JWT","groupIds":[],"attributes":[]}],"pagination":null}

curl --location  --request GET 'https://localhost:9443/internal/data/v1/applications' -kv --header 'X-WSO2-Tenant: abc.com'
{"count":2,"list":[{"id":3,"name":null,"subName":"adminfoo@abc.com","policy":"Unlimited","tokenType":"JWT","groupIds":[],"attributes":[]},{"id":4,"name":null,"subName":"adminfoo@abc.com","policy":"10PerMin","tokenType":"JWT","groupIds":[],"attributes":[]}],"pagination":null}

**/subscription-policies**
 curl --location  --request GET 'https://localhost:9443/internal/data/v1/subscription-policies'
{"count":10,"list":[{"id":1,"tenantId":-1234,"name":"Gold","quotaType":"requestCount"},{"id":2,"tenantId":-1234,"name":"Silver","quotaType":"requestCount"},{"id":3,"tenantId":-1234,"name":"Bronze","quotaType":"requestCount"},{"id":4,"tenantId":-1234,"name":"Unauthenticated","quotaType":"requestCount"},{"id":5,"tenantId":-1234,"name":"Unlimited","quotaType":"requestCount"},{"id":6,"tenantId":1,"name":"Gold","quotaType":"requestCount"},{"id":7,"tenantId":1,"name":"Silver","quotaType":"requestCount"},{"id":8,"tenantId":1,"name":"Bronze","quotaType":"requestCount"},{"id":9,"tenantId":1,"name":"Unauthenticated","quotaType":"requestCount"},{"id":10,"tenantId":1,"name":"Unlimited","quotaType":"requestCount"}],"pagination":null}

curl --location  --request GET 'https://localhost:9443/internal/data/v1/subscription-policies?policyName=Unauthenticated' -kv --header 'X-WSO2-Tenant: abc.com'
{"count":1,"list":[{"id":9,"tenantId":1,"name":"Unauthenticated","quotaType":"requestCount"}],"pagination":null}

curl --location  --request GET 'https://localhost:9443/internal/data/v1/subscription-policies' -kv --header 'X-WSO2-Tenant: abc.com'
{"count":5,"list":[{"id":6,"tenantId":1,"name":"Gold","quotaType":"requestCount"},{"id":7,"tenantId":1,"name":"Silver","quotaType":"requestCount"},{"id":8,"tenantId":1,"name":"Bronze","quotaType":"requestCount"},{"id":9,"tenantId":1,"name":"Unauthenticated","quotaType":"requestCount"},{"id":10,"tenantId":1,"name":"Unlimited","quotaType":"requestCount"}],"pagination":null}

**/application-policies**
 curl --location  --request GET 'https://localhost:9443/internal/data/v1/application-policies'
{"count":8,"list":[{"id":1,"tenantId":-1234,"name":"50PerMin","quotaType":"requestCount"},{"id":2,"tenantId":-1234,"name":"20PerMin","quotaType":"requestCount"},{"id":3,"tenantId":-1234,"name":"10PerMin","quotaType":"requestCount"},{"id":4,"tenantId":-1234,"name":"Unlimited","quotaType":"requestCount"},{"id":5,"tenantId":1,"name":"50PerMin","quotaType":"requestCount"},{"id":6,"tenantId":1,"name":"20PerMin","quotaType":"requestCount"},{"id":7,"tenantId":1,"name":"10PerMin","quotaType":"requestCount"},{"id":8,"tenantId":1,"name":"Unlimited","quotaType":"requestCount"}],"pagination":null}

curl --location  --request GET 'https://localhost:9443/internal/data/v1/application-policies?policyName=50PerMin' -kv --header 'X-WSO2-Tenant: abc.com'
{"count":1,"list":[{"id":5,"tenantId":1,"name":"50PerMin","quotaType":"requestCount"}],"pagination":null}

 curl --location  --request GET 'https://localhost:9443/internal/data/v1/application-policies' -kv --header 'X-WSO2-Tenant: abc.com'
{"count":4,"list":[{"id":5,"tenantId":1,"name":"50PerMin","quotaType":"requestCount"},{"id":6,"tenantId":1,"name":"20PerMin","quotaType":"requestCount"},{"id":7,"tenantId":1,"name":"10PerMin","quotaType":"requestCount"},{"id":8,"tenantId":1,"name":"Unlimited","quotaType":"requestCount"}],"pagination":null}

**/application-key-mappings**
curl --location  --request GET 'https://localhost:9443/internal/data/v1/application-key-mappings' -kv --header 'X-WSO2-Tenant: abc.com'
{"count":1,"list":[{"applicationId":3,"consumerKey":"43Ehg6ZSFASUsMdSSuD4XfzLnDoa","keyType":"SANDBOX"}],"pagination":null}

curl --location  --request GET 'https://localhost:9443/internal/data/v1/application-key-mappings' -k
{"count":2,"list":[{"applicationId":2,"consumerKey":"qKH901QpbwsUeaBekqhpGIrJqU4a","keyType":"PRODUCTION"},{"applicationId":3,"consumerKey":"43Ehg6ZSFASUsMdSSuD4XfzLnDoa","keyType":"SANDBOX"}],"pagination":null}

curl --location  --request GET 'https://localhost:9443/internal/data/v1/application-key-mappings?consumerKey=qKH901QpbwsUeaBekqhpGIrJqU4a' -k
{"count":1,"list":[{"applicationId":2,"consumerKey":"qKH901QpbwsUeaBekqhpGIrJqU4a","keyType":"PRODUCTION"}],"pagination":null}

//TODO
1. Add a composite API which can be used to retrieve all the data required for subscription validation at once for a given request to the gateway.
2. Maintain a state to track whether the initial data loading is done,
3. Add scopes to API resources.
4. Retrieve group Ids, attributes in the Application Resource.

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/InMemorySubscriptionValidationConstants.java [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+                    "   APP.GROUP_ID AS GROUP_ID," +
+                    "   SUB.SUBSCRIBER_ID AS SUB_ID," +
+                    "   SUB.TENANT_ID AS TENANT_ID" +
+                    " FROM " +
+                    "   AM_APPLICATION AS APP," +
+                    "   AM_SUBSCRIBER AS SUB" +
+                    " WHERE " +
+                    "   APP.SUBSCRIBER_ID = SUB.SUBSCRIBER_ID ";
+
+    public static final String SUBSCRIPTION_LOAD_SQL =
```
Comment (tharindu1st) [inline components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/InMemorySubscriptionValidationConstants.java:37]: SUBSCRIPTION_LOAD_SQL -> GET_ALL_SUBSCRIPTIONS

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/InMemorySubscriptionValidationConstants.java [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+                    "   APPLICATION_TIER," +
+                    "   TOKEN_TYPE," +
+                    "   SUBSCRIBER_ID," +
+                    "   GROUP_ID" +
+                    " FROM " +
+                    "   AM_APPLICATION" +
+                    " WHERE " +
+                    "   APPLICATION_ID = ? ";
+
+    public static final String TENANT_SUBSCRIPTION_LOAD_SQL =
```
Comment (tharindu1st) [inline components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/InMemorySubscriptionValidationConstants.java:152]: same goes here

File: features/apimgt/org.wso2.carbon.apimgt.core.feature/src/main/resources/conf_templates/templates/repository/conf/api-manager.xml.j2 [hunk 3]
```diff
@@ -363,6 +363,14 @@
        {% if apim.jwt.expiry_time is defined %}
         <JWTExpiryTime>{{apim.jwt.expiry_time}}</JWTExpiryTime>
        {% endif %}
+        {% if apim.key_manager.inmemory_sub_validation is defined %}
+        <InMemorySubscriptionValidation>
+            <Enabled>{{apim.key_manager.inmemory_sub_validation.enable}}</Enabled>
```
Comment (bhathiya) [inline features/apimgt/org.wso2.carbon.apimgt.core.feature/src/main/resources/conf_templates/templates/repository/conf/api-manager.xml.j2:368]:
```suggestion
            <Enabled>{{apim.key_manager.inmemory_subscription_validation.enabled}}</Enabled>
```

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConstants.java [hunk 4]
```diff
@@ -631,6 +634,15 @@ private Permissions() {
     public static final String API_KEY_VALIDATOR_APPLICATION_ACCESS_TOKEN_VALIDATION_PERIOD = API_KEY_VALIDATOR
             + "ApplicationTokenDefaultValidityPeriod";
 
+    public static class ApiKeyValidator {
```
Comment (tharindu1st) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConstants.java:637]: Do we need this

File: components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/model/impl/SubscriptionDataStoreImpl.java [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+//                                log.debug("Calling loadAllApiPolicies.");
+//                                return dataLoader.loadAllApiPolicies();
+//                            } catch (APIManagementException e) {
+//                                log.error("Exception while loading Api Policies");
+//                            }
+//                            return null;
+//                        });
+//
+//        executorService.scheduleAtFixedRate(apiPolicyLoadingTask, 0,
+//                this.mapBasedSubscriptionStoreConfig.getPolicyLoadingFrequency(), TimeUnit.SECONDS);
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/model/impl/SubscriptionDataStoreImpl.java:231]: do we need the commented code?

File: components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/model/impl/SubscriptionDataStoreImpl.java [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+
+    private void initializeStore() {
+
+        this.applicationKeyMappingMap = new ConcurrentHashMap<>();
+        this.applicationMap = new ConcurrentHashMap<>();
+        this.apiMap = new ConcurrentHashMap<>();
+//        this.policyMap = new ConcurrentHashMap<>();
+        this.subscriptionPolicyMap = new ConcurrentHashMap<>();
+        this.appPolicyMap = new ConcurrentHashMap<>();
+//        this.apiPolicyMap = new ConcurrentHashMap<>();
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/model/impl/SubscriptionDataStoreImpl.java:83]: can you please clean up the code?

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/SubscriptionValidationDAO.java [hunk 7]
```diff
@@ -0,0 +1,639 @@
...
+    private static Log log = LogFactory.getLog(SubscriptionValidationDAO.class);
+
+    /*
+     * This method can be used to retrieve all the APIs in the database
+     *
+     * @return {@link Map<Integer, API>}
+     * */
+    public static Map<Integer, API> getAllApis() {
+
+        Map<Integer, API> apiMap = new HashMap<>();
```
Comment (tharindu1st) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/SubscriptionValidationDAO.java:52]: Shall we make Integer-> String and bring key as context:version

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConstants.java (hunk 4 above)
Comment (isharac) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConstants.java:637]: no, will remove it

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/InMemorySubscriptionValidationConstants.java (hunk 2 above)
Comment (isharac) [reply components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/InMemorySubscriptionValidationConstants.java:152]: changed

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/InMemorySubscriptionValidationConstants.java (hunk 1 above)
Comment (isharac) [reply components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/InMemorySubscriptionValidationConstants.java:37]: changed

File: components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/model/impl/SubscriptionDataStoreImpl.java (hunk 6 above)
Comment (isharac) [reply components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/model/impl/SubscriptionDataStoreImpl.java:83]: sure.  thanks

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/SubscriptionValidationDAO.java (hunk 7 above)
Comment (isharac) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/SubscriptionValidationDAO.java:52]: will do

File: components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/model/impl/SubscriptionDataStoreImpl.java (hunk 5 above)
Comment (isharac) [reply components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/model/impl/SubscriptionDataStoreImpl.java:231]: this needs to implement.

File: components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/model/impl/SubscriptionDataLoaderImpl.java [hunk 8]
```diff
@@ -0,0 +1,497 @@
...
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+public class SubscriptionDataLoaderImpl implements SubscriptionDataLoader {
+
+    private static final Log log = LogFactory.getLog(SubscriptionDataLoaderImpl.class);
+    private static SubscriptionValidationConfig subscriptionValidationConfig;
+
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/model/impl/SubscriptionDataLoaderImpl.java:55]: Do we have to make this a static object?

File: components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/model/impl/SubscriptionDataLoaderImpl.java [hunk 9]
```diff
@@ -0,0 +1,497 @@
...
+    public static final int retrievalTimeoutInSeconds = 15;
+    public static final int retrievalRetries = 15;
+    public static final String UTF8 = "UTF-8";
+
+    public SubscriptionDataLoaderImpl() {
+
+        this.subscriptionValidationConfig = ServiceReferenceHolder.getInstance()
+                .getAPIManagerConfigurationService().getAPIManagerConfiguration()
+                .getSubscriptionValidationConfig();
+    }
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/model/impl/SubscriptionDataLoaderImpl.java:65]: If subscriptionValidationConfig is static then you would get a warning here saying "The static field  should be accessed in a static way"

File: components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/handlers/DefaultKeyValidationHandler.java [hunk 10]
```diff
@@ -261,4 +261,10 @@ public boolean validateScopes(TokenValidationContext validationContext) throws A
         }
         return true;
     }
+
+    public boolean validateSubscription(TokenValidationContext tokenValidationContext)
+            throws APIKeyMgtException {
+        //todo validate using inmemory store
+        return true;
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/handlers/DefaultKeyValidationHandler.java:268]: shall we keep the existing subscription validation logic and remove it once the in-memory validation is there

Comment (isharac) [reply components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/handlers/DefaultKeyValidationHandler.java]: sure

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/SubscriptionValidationConfig.java [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+            this.enabled = enabled;
+        }
+
+        public SubscriptionValidationConfig build() {
+
+            SubscriptionValidationConfig subscriptionValidationConfig = new SubscriptionValidationConfig();
+            subscriptionValidationConfig.serviceURL = serviceURL;
+            subscriptionValidationConfig.username = username;
+            subscriptionValidationConfig.password = password;
+            return subscriptionValidationConfig;
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/SubscriptionValidationConfig.java:92]: Seems like the enabled property is not set here

File: components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/SubscriptionDataHolder.java [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    public void unregisterTenantSubscriptionStore(int tenantId) {
+
+        subscriptionStore.remove(tenantId);
+    }
+
+    public SubscriptionDataStore getTenantSubscriptionStore(int tenantId) {
+
+        return subscriptionStore.get(tenantId);
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/SubscriptionDataHolder.java:56]: here we are retrieving using the tenant id . but we set the map using the tenant domain. need to use same way

**Comment by praminda** [comment]
Related to wso2/product-apim#8224

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/SubscriptionValidationConfig.java (hunk 11 above)
Comment (isharac) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/SubscriptionValidationConfig.java:93]: fixed

File: components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/model/impl/SubscriptionDataStoreImpl.java [hunk 13]
```diff
@@ -0,0 +1,250 @@
...
+    @Override
+    public ApplicationKeyMapping getKeyMappingByKey(String key) {
+
+        return applicationKeyMappingMap.get(key);
+    }
+
+    @Override
+    public API getApiByContextAndVersion(String context, String version) {
+
+        return null;
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/model/impl/SubscriptionDataStoreImpl.java:99]: missed to implement

File: components/apimgt/org.wso2.carbon.apimgt.internal.service/src/main/java/org/wso2/carbon/apimgt/internal/service/utils/SubscriptionValidationDataUtil.java [hunk 14]
```diff
@@ -0,0 +1,248 @@
...
+        SubscriptionPolicyListDTO subscriptionPolicyListDTO = new SubscriptionPolicyListDTO();
+        if (model != null) {
+            for (SubscriptionPolicy subscriptionPolicyModel : model) {
+                SubscriptionPolicyDTO subscriptionPolicyDTO = new SubscriptionPolicyDTO();
+                subscriptionPolicyDTO.setId(subscriptionPolicyModel.getId());
+                subscriptionPolicyDTO.setName(subscriptionPolicyModel.getName());
+                subscriptionPolicyDTO.setQuotaType(subscriptionPolicyModel.getQuotaType());
+                subscriptionPolicyDTO.setTenantId(subscriptionPolicyModel.getTenantId());
+
+                subscriptionPolicyListDTO.getList().add(subscriptionPolicyDTO);
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.internal.service/src/main/java/org/wso2/carbon/apimgt/internal/service/utils/SubscriptionValidationDataUtil.java:193]: rateLimitCount,rateLimitTimeUnit,stopOnQuotaReach are not set

File: components/apimgt/org.wso2.carbon.apimgt.internal.service/src/main/resources/api.yaml [hunk 15]
```diff
@@ -290,3 +571,201 @@ definitions:
...
+  # The URL Mapping resource
+  #-----------------------------------------------------
+  URLMapping:
+    properties:
+      authScheme:
+        type: string
+      throttlingPolicy:
+        type: string
+      httpMethod:
+        type: string
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.internal.service/src/main/resources/api.yaml:746]: Shall we include the resource and the scope as well

File: components/apimgt/org.wso2.carbon.apimgt.internal.service/src/main/java/org/wso2/carbon/apimgt/internal/service/utils/SubscriptionValidationDataUtil.java (hunk 14 above)
Comment (isharac) [reply components/apimgt/org.wso2.carbon.apimgt.internal.service/src/main/java/org/wso2/carbon/apimgt/internal/service/utils/SubscriptionValidationDataUtil.java:196]: fixed

**Pull request merged**